slides/03_workflow.qmd

```diff
@@ -1,10 +1,591 @@
+
+## Example - Step 2: Design docs
+
+1.  Describe the purpose and scope of the package
+2.  Analyse and describe the requirements in clear and simple terms ("prose")
+
+| Obligation level | Key word[^2] | Description    |
+|------------------|:-------------|----------------|
+| Duty             | shall        | "must have"    |
+| Desire           | should       | "nice to have" |
+| Intention        | will         | "optional"     |
+
+[^2]: [Rupp, C., & SOPHISTen, die. (2009). Requirements-Engineering und
+    -Management: Professionelle, iterative Anforderungsanalyse für die Praxis
+    (5. Ed.). Carl Hanser Verlag GmbH & Co. KG.](https://amzn.to/3DdwiBe)
+
+## Example - Step 2: Design docs
+
+**Purpose and Scope**
+
+The R package *simulatr* shall enable the creation of reproducible fake data.
+
+**Package Requirements**
+
+*simulatr* **shall** provide a function to generate normal distributed random
+data for two independent groups. The function **shall** allow flexible
+definition of sample size per group, mean per group, standard deviation per
+group. The reproducibility of the simulated data **shall** be ensured via an
+optional seed It **should** be possible to print the function result. A
+graphical presentation of the simulated data **will** also be possible.
+
+## Example - Step 2: Design docs
+
+::: columns
+::: {.column width="45%"}
+Useful formats / tools for design docs:
+
+-   [R Markdown](https://rmarkdown.rstudio.com)[^3] (\*.Rmd)
+-   [Quarto](https://quarto.org)^1^ (\*.qmd)
+-   [Overleaf](https://www.overleaf.com)[^4]
+-   [draw.io](http://draw.io)[^5]
+:::
+
+::: {.column width="55%"}
+::: text-align-center
+UML Diagram
+:::
+
+![](resources/simulatr.png){fig-align="right" width="95%"}
+:::
+:::
+
+[^3]: Based on
+    [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet),
+    the easy-to-use markup language
+
+[^4]: The easy to use, online, collaborative LaTeX editor
+
+[^5]: A picture is worth a thousand words
+
+## Example - Step 3: Packaging
+
+**R package programming**
+
+1.  Create basic package project (see [R Packages](02_r_packages.html) by
+    [Daniel](https://www.linkedin.com/in/danielsabanesbove/))
+2.  C&P existing R scripts (one-off scripts, prototype functions) and
+    refactor[^6] it if necessary
+3.  Create R generic functions
+4.  Document all functions
+
+[^6]: See also [Ensuring Quality](04_quality.html) by
+    [Friedrich](https://www.linkedin.com/in/pahlke)
+
+## Example - Step 3: Packaging
+
+One-off script as starting point:
+
+```{r, echo = TRUE}
+sim.data <- function(n1, n2, m1, m2, s1, s2) {
+    data.frame(
+        group = c(rep(1, n1), rep(2, n2)),
+        values = c(
+            rnorm(n = n1, mean = m1, sd = s1),
+            rnorm(n = n2, mean = m2, sd = s2)
+        )
+    )
+}
+```
+
+## Example - Step 3: Packaging
+
+Refactored script:
+
+```{r, echo = TRUE}
+getSimulatedTwoArmMeans <- function(n1, n2, mean1, mean2, sd1, sd2) {
```

Comment:
maybe `n, means, sds` as arguments and each have two elements?

Reply:
An adequate alternative, especially for defining the arguments for more than 2 groups. I think my variant is even clearer and simpler for beginners. So I'll leave it as it is.